cdap-app-fabric/src/main/java/co/cask/cdap/security/DefaultUGIProvider.java

```diff
@@ -0,0 +1,92 @@
+ */
+
+package co.cask.cdap.security;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.kerberos.SecurityUtil;
+import co.cask.cdap.common.utils.DirUtils;
+import co.cask.cdap.data2.security.ImpersonationInfo;
+import co.cask.cdap.data2.security.ImpersonationUtils;
+import co.cask.cdap.data2.security.UGIProvider;
+import com.google.inject.Inject;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+
+/**
+ * Provides a UGI by logging in with a keytab file for that user.
+ */
+public class DefaultUGIProvider implements UGIProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultUGIProvider.class);
+
+  private final CConfiguration cConf;
+  private final LocationFactory locationFactory;
+
+  @Inject
+  public DefaultUGIProvider(CConfiguration cConf, LocationFactory locationFactory) {
+    this.cConf = cConf;
+    this.locationFactory = locationFactory;
+  }
+
+  /**
+   * Resolves the {@link UserGroupInformation} for a given user, performing any keytab localization, if necessary.
+   *
+   * @return a {@link UserGroupInformation}, based upon the information configured for a particular user
+   * @throws IOException if there was any IOException during localization of the keytab
+   */
+  @Override
+  public UserGroupInformation getConfiguredUGI(ImpersonationInfo impersonationInfo) throws IOException {
+    LOG.debug("Configured impersonation info: {}", impersonationInfo);
+    File tempDir = DirUtils.createTempDir(new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                                                   cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile());
+
+    // by default, don't delete any file
+    File cleanupFile = null;
+
+    try {
+      URI keytabURI = URI.create(impersonationInfo.getKeytabURI());
+
+      File localKeytabFile;
+      if (keytabURI.getScheme() == null || "file".equals(keytabURI.getScheme())) {
+        localKeytabFile = new File(keytabURI.getPath());
+      } else {
+        localKeytabFile = ImpersonationUtils.localizeKeytab(locationFactory.create(keytabURI), tempDir);
+        // since we localized the keytab file, remove it after we're done with it
+        cleanupFile = localKeytabFile;
+      }
+
+      String expandedPrincipal = SecurityUtil.expandPrincipal(impersonationInfo.getPrincipal());
+      LOG.info("Logging in as: principal={}, keytab={}", expandedPrincipal, localKeytabFile);
+
+      // TODO: File#getPath vs File#getAbsolutePath
+      return UserGroupInformation.loginUserFromKeytabAndReturnUGI(expandedPrincipal, localKeytabFile.getPath());
+    } finally {
+      if (cleanupFile != null) {
+        if (!cleanupFile.delete()) {
+          LOG.warn("Failed to delete file: {}", cleanupFile);
```

Comment:
failure to delete the file, even if its just a temporary file, may indicate something going wrong, so leaving it as WARN